tools/c7n_gcp/c7n_gcp/query.py

```diff
@@ -161,13 +166,68 @@ def resources(self, query=None):
                     local_session(self.session_factory).get_default_project())
                 return []
             raise
-        self._cache.save(key, resources)
-        return self.filter_resources(resources)
 
     def augment(self, resources):
         return resources
 
 
+class ChildResourceManager(QueryResourceManager):
+
+    @staticmethod
+    def _get_parent_resource_info(resource_info, child_resource_type, parent_resource_type):
```

Comment:
use the child resource instead of the resource_info.. resource_info is gleaned from events which are subject to variance by event/method.. the child resource representation from the api has compatibility guarantees.

also you can avoid passing the other two parameters by making this a class method

Reply:
`resource_info` is guaranteed to contain enough information to retrieve child instance by def. So it automatically means there have to be parent identifier. I used `resource_info` because then there is no need to worry about other required arguments such as project, zone (if needed) etc. All this was achieved by prefilling parent `resource_info` with the child's one. Maybe I miss something. I can rewrite it if needed. Just let me know please.

Reply:
the type info get method which extracts resource info is specific to event retrieval. retrieving individual resources is not always going to be on that basis, it will also be on the basis of related filters and having the resource in hand. given we already have the resource here the interface is significantly cleaner using the api defined resource which have defined compatibility, rather than propogating event extracted information as a format which is bespoke.

Reply:
I see what you mean but I'm afraid we cannot go the proposed way due to the following obstacle. There are child resources having no information about parent or project so operating only on a resource type metadata and child instance won't be enough to execute parent's `get` method. 
Example.
According to the [spec](https://cloud.google.com/dns/docs/reference/v1beta2/dnsKeys#resource) DnsKeys resource has no parent id (ManagedZone) and project. But those are required to execute [ManagedZone#get](https://cloud.google.com/dns/docs/reference/v1beta2/managedZones/get). At the same time the parameters have to be in the child's `resource_info` because [DnsKeys#get](https://cloud.google.com/dns/docs/reference/v1beta2/dnsKeys/get) requires them too.

Reply:
do you have additional resource examples? I'm curious if `id` there also has the zone, I don't see a create api for those either.. and it looks like their automanaged by the platform.  In general, we don't want try to genericize on exceptions, we specialize for them with resource specific subclasses when we have concrete real world uses cases in hand.

Reply:
ie. is there actually a real use case for dns keys (ie a part of the platform the user cant modify)?

Reply:
Here is another example. AppEngine's firewall [ingress rules](https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.firewall.ingressRules/get). Although there is only one AppEngine app per project, we still will have to retrieve it to preserve consistency. And there is no app name in the response (directly or indirectly in some field).
Going back to DnsKeys. There are no composite fields (even id) with project or zone.

Reply:
> ie. is there actually a real use case for dns keys (ie a part of the platform the user cant modify)?

I can't imagine those at the moment. As far as I know they are generated when you create a zone (in case you enable DNSSEC)

Reply:
what i'm trying to understand is how common this is across gcp resources, because if its an exception, we would deal with it as a set of subclasses for that case, instead of generalizing the exception as the norm.

Reply:
Just want to let you know that i'm investigating different resources with parent-child relationship at the moment. Soon I will get back with the result.

Reply:
Haven't found anything except DnsKeys and Firewall rules mentioned above. Other resources seem to either have fields directly related to a parent or `selfLink` which can also be used to extract parent info. I will implement solution based on child instance.

Reply:
Done